libeos-parental-controls/tests/app-filter.c

```diff
@@ -236,6 +289,972 @@ test_app_filter_builder_copy_full (void)
+
+          g_dbus_method_invocation_return_value (property_invocation, NULL);
+        }
+    }
+}
+
+/* Test that setting an #EpcAppFilter on the mock D-Bus service works. The
+ * @test_data is a boolean value indicating whether to do the call
+ * synchronously (%FALSE) or asynchronously (%TRUE).
+ *
+ * The mock D-Bus replies are generated in set_app_filter_server_cb(), which is
+ * used for both synchronous and asynchronous calls. */
+static void
+test_app_filter_bus_set (BusFixture    *fixture,
+                         gconstpointer  test_data)
+{
+  gboolean success;
+  g_auto(EpcAppFilterBuilder) builder = EPC_APP_FILTER_BUILDER_INIT ();
+  g_autoptr(EpcAppFilter) app_filter = NULL;
+  g_autoptr(GError) local_error = NULL;
+  gboolean test_async = GPOINTER_TO_UINT (test_data);
+  const SetAppFilterData set_app_filter_data =
+    {
+      .expected_uid = fixture->valid_uid,
+      .expected_app_filter_value = "(false, ['/usr/bin/false', '/usr/bin/banned', 'app/org.gnome.Nasty/x86_64/stable'])",
+      .expected_oars_filter_value = "('oars-1.1', { 'violence-fantasy': 'intense' })",
+      .expected_allow_user_installation_value = "true",
+      .expected_allow_system_installation_value = "true",
+      .error_index = -1,
+    };
+
+  /* Build an app filter. */
+  epc_app_filter_builder_blacklist_path (&builder, "/usr/bin/false");
+  epc_app_filter_builder_blacklist_path (&builder, "/usr/bin/banned");
+  epc_app_filter_builder_blacklist_flatpak_ref (&builder, "app/org.gnome.Nasty/x86_64/stable");
+  epc_app_filter_builder_set_oars_value (&builder, "violence-fantasy", EPC_APP_FILTER_OARS_VALUE_INTENSE);
+  epc_app_filter_builder_set_allow_user_installation (&builder, TRUE);
+  epc_app_filter_builder_set_allow_system_installation (&builder, TRUE);
+
+  app_filter = epc_app_filter_builder_end (&builder);
+
+  /* Set the mock service function and set the filter. */
+  gt_dbus_queue_set_server_func (fixture->queue, set_app_filter_server_cb,
+                                 &set_app_filter_data);
+
+  if (test_async)
+    {
+      g_autoptr(GAsyncResult) result = NULL;
+
+      epc_set_app_filter_async (gt_dbus_queue_get_client_connection (fixture->queue),
+                                fixture->valid_uid, app_filter,
+                                FALSE, NULL, async_result_cb, &result);
+
+      while (result == NULL)
+        g_main_context_iteration (NULL, TRUE);
+      success = epc_set_app_filter_finish (result, &local_error);
+    }
+  else
+    {
+      success = epc_set_app_filter (gt_dbus_queue_get_client_connection (fixture->queue),
+                                    fixture->valid_uid, app_filter,
+                                    FALSE, NULL, &local_error);
+    }
+
+  g_assert_no_error (local_error);
+  g_assert_true (success);
+}
+
+/* Test that epc_set_app_filter() returns an appropriate error if the mock D-Bus
+ * service reports that the given user cannot be found.
+ *
+ * The mock D-Bus replies are generated inline. */
```

Comment:
What's the rationale for doing some of these tests in the "worker thread" style, and some in this "async + inline replies" style? I guess it does demonstrate both models of GtDBusQueue usage.

Reply:
The worker thread style is needed whenever a test uses the synchronous libeos-parental-controls API. Any other test which can sensibly reuse the same worker thread function also uses it. Everything else is done in the ‘inline replies’ style because it keeps the client/server conversation flow in a single function, so it should be a bit more obvious what’s happening.

And yes, it is partially to demonstrate both models of `GtDBusQueue` usage.

Does that make sense?

Reply:
Yep!